qt/aqt/sidebar.py

```diff
@@ -1078,47 +1246,110 @@ def _get_saved_searches(self) -> Dict[str, str]:
+        ):
             return
-        conf[new] = filt
-        del conf[old]
+        conf[new_name] = filt
+        del conf[old_name]
         self._set_saved_searches(conf)
-        self.refresh()
+        self.refresh(
+            lambda item: item.item_type == SidebarItemType.SAVED_SEARCH
+            and item.name == new_name
+        )
 
-    def save_current_search(self, _item: Any = None) -> None:
+    def save_current_search(self) -> None:
         try:
             filt = self.col.build_search_string(
                 self.browser.form.searchEdit.lineEdit().text()
             )
         except InvalidInput as e:
             show_invalid_search_error(e)
-        else:
-            name = getOnlyText(tr(TR.BROWSING_PLEASE_GIVE_YOUR_FILTER_A_NAME))
-            if not name:
-                return
-            conf = self._get_saved_searches()
-            conf[name] = filt
-            self._set_saved_searches(conf)
-            self.refresh()
+            return
+        name = getOnlyText(tr(TR.BROWSING_PLEASE_GIVE_YOUR_FILTER_A_NAME))
+        if not name:
+            return
+        conf = self._get_saved_searches()
+        if name in conf and not askUser(
+            tr(TR.BROWSING_CONFIRM_SAVED_SEARCH_OVERWRITE, name=name)
+        ):
+            return
+        conf[name] = filt
+        self._set_saved_searches(conf)
+        self.refresh(
+            lambda item: item.item_type == SidebarItemType.SAVED_SEARCH
+            and item.name == name
+        )
 
     def manage_notetype(self, item: SidebarItem) -> None:
         Models(
             self.mw, parent=self.browser, fromMain=True, selected_notetype_id=item.id
         )
+
+    def manage_template(self, item: SidebarItem) -> None:
+        note = Note(self.col, self.col.models.get(item._parent_item.id))
+        CardLayout(self.mw, note, ord=item.id, parent=self, fill_empty=True)
+
+    # Helpers
+    ##################
+
+    def _selected_items(self) -> List[SidebarItem]:
+        return [self.model().item_for_index(idx) for idx in self.selectedIndexes()]
+
+    def _selected_decks(self) -> List[int]:
+        return [
+            item.id
+            for item in self._selected_items()
+            if item.item_type == SidebarItemType.DECK
+        ]
+
+    def _selected_saved_searches(self) -> List[str]:
+        return [
+            item.name
+            for item in self._selected_items()
+            if item.item_type == SidebarItemType.SAVED_SEARCH
+        ]
+
+    def _selected_tags(self) -> List[str]:
+        return [
+            item.full_name
+            for item in self._selected_items()
+            if item.item_type == SidebarItemType.TAG
+        ]
+
+    def ask_remove_tags(self, tags: List[str]) -> bool:
+        count = len(
+            self.col.find_notes(
+                self.col.build_search_string(
+                    *(SearchNode(tag=tag) for tag in tags), joiner="OR"
+                )
+            )
+        )
```

Comment:
Counting the number of notes matching the tag can be fairly slow on a large collection, especially if the user has a spinning disk. In the deck deletion case, we display the number of cards up front, because it may not be clear to the user that deleting the deck will result in cards (and also notes) being deleted, and also because they're comparatively cheap to count. But in this case there's no risk of anything but the tag being deleted - so I wonder if we need to be presenting the count to the user up front?

If you do think we should keep it, it would be better run in the background (eg via mw.taskman.run_in_background()), so that the UI doesn't freeze until the data becomes available. 

I pushed the initial work on multi-step undo support earlier today, and as that gets enabled for more of the codebase, it will allow us to be less conservative with confirmation prompts: https://goodui.org/#8

Reply:
Apart from asking for confirmation if any notes were to be affected, my intention was to indicate to the user that this operation may take a while if a lot of notes needed to be updated. But I see now that it might do more harm than good.

Favouring reversability over confirmation sounds very reasonable, although undoing will only help you if you realise your mistake and if you delete a deck thinking it didn't contain any notes or a tag thinking it wasn't used, you might not find out until much later.
The logical consequence seems to be to remove the confirmation prompt but show the same information afterwards as in the form of an uninstrusive tooltip. The counts can then be retrieved from the backend without further cost.
What do you reckon?

Reply:
Yeah, I think reporting the affected count at the end of the operation is best, as it avoids all the duplicate work. Some of the other backend methods like tags.bulk_add() and tags.rename() already do this.

For the operations that we think might take a while, we should probably be using taskman.with_progress() instead of run_in_background(), as that'll take care of showing a progress dialog as well.

Reply:
I switched to `with_progress()`, but for the record: I see a progress dialogue either way and for some operations it freezes together with the  GUI.

Reply:
Which op is freezing? The progress window usually should pop up after a few seconds IIRC.

Reply:
It pops up but then everything freezes altogether. It doesn't seem to depend on the operation. Sometimes it works, but more often than not, the progress animation is frozen and Windows marks the application as unresponsive.

![2PgqFliBWA](https://user-images.githubusercontent.com/48286681/110781310-0cebff00-8266-11eb-81fa-54be62906e75.gif)

Reply:
Thanks for the repro, I've posted an update on https://github.com/ankitects/anki/issues/1065